sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala

```diff
@@ -654,6 +654,45 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-29166: dynamic partition overwrite with limitation") {
+    withSQLConf(
+      SQLConf.PARTITION_OVERWRITE_MODE.key -> PartitionOverwriteMode.DYNAMIC.toString) {
+      withTable("t") {
+        sql(
+          """
+            |create table t(i int, part1 int, part2 int) using parquet
+            |partitioned by (part1, part2)
+          """.stripMargin)
+
+        // no restriction
+        sql("insert overwrite table t partition(part1=2, part2)" +
+          " select 2, explode(array(2, 3, 4, 5))")
+        checkAnswer(spark.table("t"),
```

Comment:
What i meant was something like this (each dynamic partition insert is a separate query):
sql("insert overwrite table t partition(part1=2, part2)" +
              " select 2, 2)") //this will create say the 1st partition
sql("insert overwrite table t partition(part1=2, part2)" +
              " select 2, 3")//this will create say the 2nd partition
sql("insert overwrite table t partition(part1=2, part2)" +
              " select 2, 4)")//this will create say the 3rd partition
sql("insert overwrite table t partition(part1=2, part2)" +
              " select 2, 5")//this will create say the 4th partition

Now when do the last sql(trying to create the 4th partition), what is the behaviour in HIVE and what is expected in spark with your change when we set SQLConf.DYNAMIC_PARTITION_MAX_PARTITIONS.key -> "3"?

Reply:
The limitation parameter won't act on the SQLs above, it doesn't restrict how many total partitions a table can be inserted whatever dynamic or not. They are 4 normal queries. It only to limit some unexpected insert like "insert overwrite table t partition(part1=2, part2) select 2, id" when "id" is a mistake partition column.  I think the behaviour in Hive is same with here, but I will try this in Hive.